### PR TITLE
[canvaskit] Handle case where platform view is prerolled but not composited

### PR DESCRIPTION
**Pull request by hterkelsen**
Fixes https://github.com/flutter/flutter/issues/88546

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on
writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/master/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: lib/web_ui/lib/src/engine/canvaskit/embedded_views.dart [hunk 1]
```diff
@@ -93,47 +100,63 @@ class HtmlViewEmbedder {
     _frameSize = size;
   }
 
-  List<CkCanvas> getCurrentCanvases() {
+  /// Returns a list of canvases which will be overlaid on top of the "base"
+  /// canvas after a platform view is composited into the scene.
+  ///
+  /// The engine asks for the overlay canvases immediately before the paint
+  /// phase, after the preroll phase. Note that in the preroll phase we must
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/canvaskit/embedded_views.dart:107]: nit: our style guide doesn't like "Note that" https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#avoid-empty-prose

File: lib/web_ui/lib/src/engine/canvaskit/embedded_views.dart [hunk 2]
```diff
@@ -145,6 +168,16 @@ class HtmlViewEmbedder {
   /// If this returns a [CkCanvas], then that canvas should be the new leaf
   /// node. Otherwise, keep the same leaf node.
   CkCanvas? compositeEmbeddedView(int viewId) {
+    final int numCompositedViews = _compositionOrder.length;
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/canvaskit/embedded_views.dart:171]: nit: consider `compositedViewCount` (https://dart.dev/guides/language/effective-dart/design#avoid-abbreviations)

File: lib/web_ui/lib/src/engine/canvaskit/embedded_views.dart [hunk 3]
```diff
@@ -414,6 +447,7 @@ class HtmlViewEmbedder {
         }
       }
     }
+    _pictureRecordersCreatedDuringPreroll.clear();
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/canvaskit/embedded_views.dart:450]: Do we need to call `PictureRecorder.delete()` somewhere? If so, do we?

File: lib/web_ui/test/canvaskit/embedded_views_test.dart [hunk 4]
```diff
@@ -554,6 +554,26 @@ void testMain() {
...
+      await _createPlatformView(0, 'test-platform-view');
+
+      final EnginePlatformDispatcher dispatcher =
+          ui.window.platformDispatcher as EnginePlatformDispatcher;
+
+      LayerSceneBuilder sb = LayerSceneBuilder();
+      sb.pushOffset(0, 0);
+      sb.pushClipRect(ui.Rect.zero);
+      sb.addPlatformView(0, width: 10, height: 10);
+      sb.pop();
```
Comment (yjbanov) [inline lib/web_ui/test/canvaskit/embedded_views_test.dart:573]: Checking the DOM for the absence of `<flt-platform-view-slot>` might help avoid false positives (although the check itself must not be prone to false negatives :thinking:).

**Review by yjbanov — APPROVED** [review]
lgtm

File: lib/web_ui/lib/src/engine/canvaskit/embedded_views.dart (hunk 3 above)
Comment (hterkelsen) [reply lib/web_ui/lib/src/engine/canvaskit/embedded_views.dart:450]: They are deleted when `endRecording` is called (in the above code)

File: lib/web_ui/test/canvaskit/embedded_views_test.dart (hunk 4 above)
Comment (hterkelsen) [reply lib/web_ui/test/canvaskit/embedded_views_test.dart:573]: Done

File: lib/web_ui/lib/src/engine/canvaskit/embedded_views.dart (hunk 2 above)
Comment (hterkelsen) [reply lib/web_ui/lib/src/engine/canvaskit/embedded_views.dart]: Done

File: lib/web_ui/lib/src/engine/canvaskit/embedded_views.dart (hunk 1 above)
Comment (hterkelsen) [reply lib/web_ui/lib/src/engine/canvaskit/embedded_views.dart]: Done

**Pull request merged**